### ISSUE TITLE
Allow DbContext.Update to delete related entities

### ISSUE THREAD

**Issue by Dunge** [dotnet/efcore#17626]
I'm still on 2.2 due to framework dependencies, but I don't think this behave differently in 3.0.

DbContext.Update (and DbSet.Update) seems to be a godsend in a disconnected scenario. You just open the context, call Update passing the new entity and it will automatically do lots of operations for you, including looking if it's a new one (create or update) and doing a recursive search of the navigation properties to update all related entities too. By related entities I mean relationships declared on the model with "HasOne/HasMany".

If you pass an entity containing a related entity with an existing ID, it will update its properties correctly. If you pass a new related entity with an ID of 0, it will be created correctly. But in the case where you pass an empty list or a list with one less entity in it than there is in the database, or null for the "HasOne" relationship, it WON'T delete it. This is documented [here](https://docs.microsoft.com/en-us/ef/core/saving/disconnected-entities#handling-deletes) in the section "Handling deletes". _Delete can be tricky to handle since often the absence of an entity means that it should be deleted._ 

The solution proposed in the documentation linked above suggest doing a "graph diff". This requires you to drop the use of Update() completely and do what it does manually instead: load the existing entity and use CurrentValues.SetValues() for the properties, then compare each related entities ID and call Add/Remove manually. This code works for a small model, but is awfully hard to implement in a generic manner for a huge model with nested related types and inheritance.

A similar issue #9312 was once opened in 2017 (back in EFCore1). The important argument was that EFCore doesn't know if the entity is "owned" or simply referenced, so it doesn't know if it should delete it or not. At this point, the suggestion was to wait for "OwnsOne/OwnsMany" to be implemented in 2.0 for this. After a few tests, I can confirm that defining entities with OwnsMany doesn't handle delete differently than defining them with HasMany. The only difference I saw is that I didn't need to call "Include" in my queries anymore to load them. Unfortunately, I've read here that starting from 3.0 it also brings the disadvantage that you can't query these entities as a DbSet independently from the owner anymore.

With the "Cascade Delete" rule, deleting the base entity correctly loop through all related data and also delete them correctly, why doesn't Update behave the same? I'm aware that cascade delete is done by the SQL server itself and if defined in the table schema, while what I'm looking for require to launch a "DELETE" query.

I wouldn't suggest an additional parameter to DbContext.Update() since it could differ depending on which related data in the same entity, but I don't see why we can't have an additional parameter in the model creation "HasOne/HasMany" calls, or maybe an "OnUpdate" method just like we do with .OnDelete(DeleteBehavior.Cascade), or just do it by default with the "Owns" definitions.

I tried to implement a solution by having all my entities implementing an interface that force them to implement a function that would returns a list of all the collections to update, which would in turn execute the code in the documentation for each, but failed miserably. I also looked at the TrackGraph but it doesn't help. If anyone have a clean generic solution I'm all ear. I also asked a stackoverflow question [here](https://stackoverflow.com/questions/57400359/entity-framework-core-saving-disconnected-entities-and-handling-delete-on-relate). Plus, doing it manually seems counter efficient since Entity Framework already knows everything it requires (the keys and the relationships) based on the model.

### Further technical details
EF Core version: 2.2.6
Database Provider: Microsoft.EntityFrameworkCore.SqlServer 2.2.6
Operating system: Windows 10 1809
IDE: Visual Studio 2017 15.9.14

**Comment by ajcvickers** [comment]
@Dunge I think what you are looking for is basically #5536, or some part of that. The difficulty with deletes is knowing which entities to delete. This requires one of:
* Getting back from the client explicitly which entities in the collection were deleted
* Getting the entities from the database such that they can be compared with the entities sent back from the client
* Treating the graph of as an aggregate such that the entire collection is replaced every time. This isn't very effective/efficient if the graph is not conceptually an aggregate--i.e. it wouldn't be a great workaround for a normal one-to-many relationship.

So I'm closing this as a duplicate of #5536, although #4025 would also provide a nice solution to this problem if the client just marked the entities as soft-deleted and EF did the rest. /cc @divega

**Comment by divega** [comment]
@ajcvickers, @AndriySvyryd however this part concerns me:

> After a few tests, I can confirm that defining entities with OwnsMany doesn't handle delete differently than defining them with HasMany. The only difference I saw is that I didn't need to call "Include" in my queries anymore to load them. 

I think if this is true (I am not going to test it right now), we may be missing an important part of ownership. 

I will reopen this just so that we can discuss it in triage on our next chance.

**Comment by AndriySvyryd** [comment]
>>After a few tests, I can confirm that defining entities with OwnsMany doesn't handle delete differently than defining them with HasMany. The only difference I saw is that I didn't need to call "Include" in my queries anymore to load them.

>I think if this is true (I am not going to test it right now), we may be missing an important part of ownership.

This is tracked by https://github.com/aspnet/EntityFrameworkCore/issues/10551 and by https://github.com/aspnet/EntityFrameworkCore/issues/1985

**Comment by Dunge** [comment]
Thanks for reviewing and the amazing moderation work you guys do. I'm surprised I haven't seen those others issues when I searched before creating mine, so sorry for the pollution.

@divega as I said, this is what I observed on 2.2, I didn't try on 3.0

I feel like aggregate / owned type / hard relationship / composite (no matter how you call it) versus a soft link / reference is a concept that is well used in data modeling, I'm surprised this is not massively used and easier to work with under EF. How else are everyone else handling saving most of their data exactly? Am I missing something obvious?

Some comments about what @ajcvickers said:
> Getting back from the client explicitly which entities in the collection were deleted

I've seen this proposed as a solution in a few articles. This is viable, but it add a lot of complexity to the operation, especially when the entities transfers different serialization layers. Instead of just transferring the POCO instance that is the raw element to save, you also have to transfer some kind of change tracking graph with relations to delete? What if the client don't want know about the previous state of the entity? What if he tampers with the list of relations to delete and omit some, couldn't that lead to orphaned rows in the database?

> Getting the entities from the database such that they can be compared with the entities sent back from the client

Sure, this is what the solution proposed in MSDN do. In my situation I also need to read the entities from the database to build a changeset anyway, I think for many situations this is required.

> Treating the graph of as an aggregate such that the entire collection is replaced every time. This isn't very effective/efficient if the graph is not conceptually an aggregate--i.e. it wouldn't be a great workaround for a normal one-to-many relationship.

As previously mentioned, that's why relationships need to be marked in one way of another to differentiate the type of relationships. Defining it in the model as "owned" should take care of this.

> #4025 would also provide a nice solution to this problem if the client just marked the entities as soft-deleted and EF did the rest.

Soft delete can be great in certain scenarios, but sometimes data needs to be deleted for real.


In any case, I will update my StackOverflow question with a cheap solution for working around this with EF 2.0 for anyone else stumbling on this issue until there's an official way. It worked for my use case even if the code is not the cleanest.
